Assets/Scripts/Layers/Adapters/ChainOfResponsibility.cs

```diff
@@ -0,0 +1,134 @@
+        {
+            if(chain != null)
+                StopCoroutine(chain);
+
+            chain = StartCoroutine(DownloadAndCheckSupport(TargetUrl));
+        }
+
+        private IEnumerator DownloadAndCheckSupport(string url)
+        {
+            // Start by download the file, so we can do a detailed check of the content to determine the type
+            var urlAndData = new LocalFile()
+            {
+                SourceUrl = url,
+                LocalFilePath = ""
+            };
+            yield return DownloadDataToLocalCache(urlAndData);
+
+            // No local cache? Download failed.
+            if(string.IsNullOrEmpty(urlAndData.LocalFilePath))
+            {
+                OnDownloadFailed.Invoke(url);
+                yield break;
+            }
+
+            // Find the proper adapter in a chain of responsibility
+            yield return AdapterChain(urlAndData);
+        }
+
+        /// <summary>
+        /// We download the file to a local cache so we can check the content from the adapters using streamreading.
+        /// This way we keep the heap memory usage low and can handle large files (like large obj's, or large WFS responses)
+        /// </summary>
+        /// <param name="urlAndData">The local file object where the path will set</param>
+        /// <returns></returns>
+        private IEnumerator DownloadDataToLocalCache(LocalFile urlAndData)
+        {
+            var url = urlAndData.SourceUrl;
+            var optionalExtention = Path.GetExtension(url);
+            var uwr = UnityWebRequest.Get(url);
+            var guidFilename = Guid.NewGuid().ToString() + optionalExtention;
+            string path = Path.Combine(Application.persistentDataPath, guidFilename);
+
+            uwr.downloadHandler = new DownloadHandlerFile(path);
+            yield return uwr.SendWebRequest();
+            if (uwr.result == UnityWebRequest.Result.Success)
+            {
+                urlAndData.LocalFilePath = path;
+            }
+            else
+            {
+                urlAndData.LocalFilePath = "";
+                Debug.LogError("Download failed: " + uwr.error);
+            }
+        }
+
+        private IEnumerator AdapterChain(LocalFile urlAndData)
+        {
+            // Get our interface references
+            dataTypeAdapterInterfaces = new IDataTypeAdapter[dataTypeAdapters.Length];
+            for (int i = 0; i < dataTypeAdapters.Length; i++)
+                dataTypeAdapterInterfaces[i] = dataTypeAdapters[i] as IDataTypeAdapter;
+
+            // Check data type per adapter using order set in inspector
+            foreach (var adapter in dataTypeAdapterInterfaces)
+            {
+                if (adapter.Supports(urlAndData))
+                {
+                    Debug.Log("<color=green>Adapter found: " + adapter.GetType().Name + "</color>");
+                    adapter.Execute(urlAndData);
+                    OnAdapterFound.Invoke();
```

Comment:
Would it be interesting to pass the adapter here as well?

Reply:
yes, nice